---
title: |
    | Targeted Sequencing Summary Report: 
    | `r config$Desc_Name`
date: |
    | report generated : `r Sys.Date()`
    | software version : `r soft_version`
    | environment build : `r build_version`
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: 
      collapsed: false
    number_sections: false
    fig_caption: true
    theme: cerulean 
    highlight: tango
    smart: false
  pdf_document:
    toc: true
    toc_depth: 2
    number_sections: false
    fig_caption: true
header-includes: 
  - \usepackage{float}
  - \usepackage{indentfirst}
  - \usepackage{booktabs}
  - \usepackage{longtable}
  - \usepackage{gensymb}
  - \setlength{\defaultaddspace}{0em}
  - \setlength{\parindent}{2em}
fontsize: 11pt
geometry: margin=0.6in
---

```{r setup, include=FALSE}
packs <- c(
  "magrittr", "data.table", "Biostrings", "GenomicRanges",
  "knitr", "scales", "RColorBrewer", "tidyverse", "pander")

packs_loaded <- suppressMessages(sapply(packs, require, character.only = TRUE))
if(!all(packs_loaded)){
  print(data.frame(
    "R-Packages" = names(packs_loaded), 
    "Loaded" = packs_loaded, 
    row.names = NULL), row.names = FALSE)
  stop("Check dependancies.")
}

options(
  stringsAsFactors = FALSE, 
  scipen = 99, 
  knitr.table.format = "latex"
)

knitr::opts_chunk$set(
  echo = FALSE,
  comment = "",
  warning = FALSE,
  error = TRUE,
  message = FALSE,
  cache = FALSE,
  results = "asis",
  fig.align = "center",
  dpi = 300,
  dev = c("png", "pdf"),
  fig.pos = "H",
  fig.width = 7
)

if(args$figures != FALSE){
  knitr::opts_chunk$set(
    fig.path = file.path(
      output_dir, gsub("[\\w]+$", "figures/", output_file, perl = TRUE)))
}


# Custom theme
custom_theme <- theme_bw() +
  theme(
    panel.background = element_rect(color = "white", fill = "white"),
    panel.border = element_rect(color = "white"),
    panel.grid.major = element_line(color = "white"),
    panel.grid.minor = element_line(color = "white"),
    strip.background = element_rect(color = "black", fill = "white"),
    strip.text = element_text(face = "bold"),
    legend.key = element_rect(fill = "white"),
    title = element_text(face = "bold"),
    axis.ticks = element_line(color = "black"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(face = "bold", angle = 0),
    axis.title = element_text(color = "black", face = "bold"),
    axis.line = element_line(color = "black"),
    axis.line.x = element_line(color = "black"),
    axis.line.y = element_line(color = "black"))

blank_theme <- theme_bw() +
  theme(
    panel.background = element_blank(),
    panel.border = element_blank(),
    axis.line = element_blank(),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_blank(),
    strip.background = element_blank(),
    strip.text = element_text(face = "bold"),
    panel.grid = element_blank(),
    legend.position = "none",
    title = element_text(face = "bold"))
  

# Captions
tbl_caps <- c(
  "Specimen metadata.",
  "Specimen summary.",
  "On-target editing **read counts** and proportions.",
  "Off-target editing **read counts** and proportions.",
  "Control target editing **read counts** and proportions.")

fig_caps <- c(
  "Proportional on-target editing.",
  "On-target deletion profile.",
  "On-target insertion profile.",
  "Proportional off-target editing.",
  "Off-target deletion profile.",
  "Off-target insertion profile.",
  "Proportional null or control target editing.",
  "Control target deletion profile.",
  "Control target insertion profile.")

if(args$format == "html"){
  tbl_caps <- paste0("Table ", 1:5, ". ", tbl_caps)
  fig_caps <- paste0("Figure ", 1:3, ". ", fig_caps)
}

shrt_tbl <- distinct(input_uniq, specimen, edit)
```

***

# Summary

Targeted amplicon sequencing was conducted on specimens to validate nuclease editing at on- and off-target genomic loci. Genomic DNA was amplified with pairs of primers flanking `r unname(table(panel_targets$edit)["on"])` on-target(s), `r unname(table(panel_targets$edit)["off"])` off-target(s), and `r unname(table(panel_targets$edit)["null"])` control target(s) sites. Amplified product was then sequenced on an Illumina instrument and analyzed using custom-built software. In this report, results for on-, off-, and control sites are found in different sections, which all include the proportion of complete (unedited) amplicons, as well as insertion and deletion containing amplicons.

```{r} 
if(args$format == "pdf"){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

# Specimen overview

Editing conditions related for each specimen are found in Table 1. Specimen names ending with an "S" or an "M" indicate whether the specimen was processed in a single-plex fashion (each targeted amplification was conducted individually) or with a multiplexed panel (all reactions targeted at the same time, see Methods section for additional detail).

```{r spec_supp, eval=!is.null(config$Supplemental_Info)}
pander(supp_info, caption = tbl_caps[1], justify = "centre", style = "simple")
```

Total read counts corresponding to on-, off-, and control target sites can be found in Table 2. The replicate counts for each specimen indicate the number of technical replicates that were performed. InDel (or insertion / deletion) associated reads indicate the number of edited reads observed across all sites. On-, Off-, and Control target site read counts indicate the number of reads associated with each type of target site (including both complete or unedited reads and indel containing reads).

```{r spec_summary}
pander(
  summary_tbl, caption = tbl_caps[2], 
  justify = c("centre", rep("right", 6)), big.mark = ",", 
  style = "multiline", missing = 0, keep.line.breaks = TRUE)
```


```{r} 
if(args$format == "pdf"){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```


# On-target analysis

Table 3 tabulates the read counts associated with each on-target site and the associated proportion for Uncharacterized (Unchar.), Complete (unedited), Insertion, and Deletion containing reads. Read counts are in **bold**, while proportions are in plain text and between 0 and 1.

```{r on_tar_summary_tbl}
on_tar_summary_print <- on_tar_summary_tbl %>%
  group_by(specimen, rname) %>%
  mutate(
    "Reads" = pNums(total.cnt, digits = 0),
    "Unchar." = pNums(unc.freq),
    "Complete" = pNums(com.freq),
    "Insertion" = pNums(in.freq),
    "Deletion" = pNums(del.freq)) %>%
  ungroup() %>%
  select(specimen, rname, Reads, Unchar., Complete, Insertion, Deletion) %>%
  gather(Type, val, -specimen, -rname) %>%
  spread(rname, val, fill = "-") %>%
  mutate(
    specimen = factor(
      specimen, 
      levels = unique(str_extract(sample_info$sampleName, "[\\w]+"))),
    Type = factor(
      Type , 
      levels = c("Reads", "Unchar.", "Complete", "Insertion", "Deletion"))) %>%
  arrange(specimen, Type) %>%
  rename(specimen = "Specimen") %>%
  mutate(Specimen = ifelse(
    duplicated(as.character(Specimen)), " ", as.character(Specimen)))

emphasize.strong.rows(which(!grepl(" ", on_tar_summary_print$Specimen)))
pander(
  on_tar_summary_print, caption = tbl_caps[3], 
  justify = "center", style = "simple", 
  col.names = pandoc.strong.return(names(on_tar_summary_print)))

```

```{r} 
if(args$format == "pdf"){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

Similar to Table 3, Figure 1 displays the proportions associated with each on-target editing for each sample. Classes of reads are the same as previously described.

```{r on_tar_graphic, fig.width=7.5, fig.height=3.5, fig.cap=fig_caps[1]}
on_tar_plot_data <- on_tar_summary_tbl %>%
  select(-total.cnt) %>%
  gather(key = type, value = freq, -specimen, -rname) %>%
  mutate(
    rname = factor(
      rname, levels = panel_targets$target[panel_targets$edit == "on"]),
    specimen = factor(
      specimen, levels = unique(str_extract(sample_info$sampleName, "[\\w]+"))),
    type = factor(
      type, levels = c("unc.freq", "com.freq", "in.freq", "del.freq"))) %>%
  arrange(desc(type))

levels(on_tar_plot_data$type) <- c("Unchar.", "Complete", "Insertion", "Deletion")
  
ggplot(on_tar_plot_data, aes(x = rname, y = freq, group = specimen)) + 
  geom_bar(stat = "identity", aes(fill = type)) + 
  facet_wrap(~ specimen, nrow = 1) +
  scale_fill_brewer(type = "qual", palette = "Paired") +
  labs(x = NULL, y = "Frequency", fill = NULL) +
  custom_theme + 
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 45, hjust = 1))
```

Indel profiles, or the distribution of deleted DNA around the target site, are show in Figure 2 for reads containing deletions at the target sites. A different color has been given to each target to distinguish each. The height of each bar is associated with the number of reads observed missing the respective position. Deletion profiles have been separated based on treatment conditions, and samples having the same condition have their data pooled for this analysis.

```{r on_tar_del_profile, fig.width=7.5, fig.height=4, fig.cap=fig_caps[2]}
on_del_profile <- edit_grl$D[edit_grl$D$edit == "on"]
on_del_profile <- on_del_profile[
  rep(seq_along(on_del_profile), on_del_profile$count)]

plot_indel_cov(on_del_profile, norm_targets, resolution = 1L) + 
  scale_y_continuous(trans = "log10", labels = comma) + 
  scale_fill_brewer(type = "qual", palette = "Set1") + 
  custom_theme + 
  theme(legend.position = "None")
```

```{r on_tar_in_profile, eval=FALSE, fig.width=7.5, fig.height=3.5, fig.cap=fig_caps[3]}
on_in_profile <- edit_grl$I[edit_grl$I$edit == "on"]
start(on_in_profile) <- start(on_in_profile) - 
  as.integer(on_in_profile$in.len/2)
width(on_in_profile) <- on_in_profile$in.len
on_in_profile <- on_in_profile[
  rep(seq_along(on_in_profile), on_in_profile$count)]

plot_indel_cov(on_in_profile, norm_targets, resolution = 1L) + 
  scale_y_continuous(trans = "log10", labels = comma) + 
  scale_fill_brewer(type = "qual", palette = "Set1") + 
  custom_theme + 
  theme(legend.position = "None")
```


```{r} 
if(args$format == "pdf"){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```


# Off-target analysis

Similar to the on-target analysis, the off-target analysis contains the same types of tables and figures. Table 4 indicates the read counts in **bold** for each specimen and off-target site. Proportions below in plain text indicate the uncharacterized, complete, insertion containing, or deletion containing sequences.

```{r off_tar_summary_tbl}
off_tar_summary_print <- off_tar_summary_tbl %>%
  group_by(specimen, rname) %>%
  mutate(
    "Reads" = pNums(total.cnt, digits = 0),
    "Unchar." = pNums(unc.freq),
    "Complete" = pNums(com.freq),
    "Insertion" = pNums(in.freq),
    "Deletion" = pNums(del.freq)) %>%
  ungroup() %>%
  select(specimen, rname, Reads, Unchar., Complete, Insertion, Deletion) %>%
  gather(Type, val, -specimen, -rname) %>%
  spread(rname, val, fill = "-") %>%
  mutate(
    specimen = factor(
      specimen, 
      levels = unique(str_extract(sample_info$sampleName, "[\\w]+"))),
    Type = factor(
      Type , 
      levels = c("Reads", "Unchar.", "Complete", "Insertion", "Deletion"))) %>%
  arrange(specimen, Type) %>%
  rename(specimen = "Specimen") %>%
  mutate(Specimen = ifelse(
    duplicated(as.character(Specimen)), " ", as.character(Specimen)))

emphasize.strong.rows(which(!grepl(" ", off_tar_summary_print$Specimen)))
pander(
  off_tar_summary_print, caption = tbl_caps[4], 
  justify = "center", style = "simple", 
  col.names = pandoc.strong.return(names(off_tar_summary_print)))

```

```{r} 
if(args$format == "pdf"){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```

In a similar fashion to Figures 1 and 2, Figures 3 and 4 display the frequency of different types of reads across samples and off-target sites and the distribution of deleted sequences at the off target sites.

```{r off_tar_graphic, fig.width=7.5, fig.height=3.5, fig.cap=fig_caps[4]}
off_tar_plot_data <- off_tar_summary_tbl %>%
  select(-total.cnt) %>%
  gather(key = type, value = freq, -specimen, -rname) %>%
  mutate(
    rname = factor(
      rname, levels = panel_targets$target[panel_targets$edit == "off"]),
    specimen = factor(
      specimen, levels = unique(str_extract(sample_info$sampleName, "[\\w]+"))),
    type = factor(
      type, levels = c("unc.freq", "com.freq", "in.freq", "del.freq"))) %>%
  arrange(desc(type))

levels(off_tar_plot_data$type) <- c("Unchar.", "Complete", "Insertion", "Deletion")
  
ggplot(off_tar_plot_data, aes(x = rname, y = freq, group = specimen)) + 
  geom_bar(stat = "identity", aes(fill = type)) + 
  facet_wrap(~ specimen, nrow = 1) +
  scale_fill_brewer(type = "qual", palette = "Paired") +
  labs(x = NULL, y = "Frequency", fill = NULL) +
  custom_theme +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r off_tar_del_profile, fig.width=7.5, fig.height=3.5, fig.cap=fig_caps[5]}
off_del_profile <- edit_grl$D[edit_grl$D$edit == "off"]
off_del_profile <- off_del_profile[
  rep(seq_along(off_del_profile), off_del_profile$count)]

plot_indel_cov(off_del_profile, norm_targets, resolution = 1L) + 
  scale_y_continuous(trans = "log10", labels = comma) + 
  scale_fill_brewer(type = "qual", palette = "Dark2") + 
  custom_theme + 
  theme(legend.position = "None")
```

```{r off_tar_in_profile, eval=FALSE, fig.width=7.5, fig.height=3.5, fig.cap=fig_caps[6]}
off_in_profile <- edit_grl$I[edit_grl$I$edit == "off"]
start(off_in_profile) <- start(off_in_profile) - 
  as.integer(off_in_profile$in.len/2)
width(off_in_profile) <- off_in_profile$in.len
off_in_profile <- off_in_profile[
  rep(seq_along(off_in_profile), off_in_profile$count)]

plot_indel_cov(off_in_profile, norm_targets, resolution = 1L) + 
  scale_y_continuous(trans = "log10", labels = comma) + 
  scale_fill_brewer(type = "qual", palette = "Set1", direction = -1) + 
  custom_theme + 
  theme(legend.position = "None")
```

```{r} 
if(args$format == "pdf"){
  cat("\\newpage")
}else{
  cat('<P style="page-break-before: always">')
}
```


# Control-target analysis

As with Tables 2 and 3, Table 4 tabulates the read counts and proportions for control sites, or sites included in the panel that should not display any background. Hence, editing observed at these sites could be interpreted as noise within the analysis. Figure 5 displays the proportional information in a barplot.

```{r null_tar_summary_tbl}
null_tar_summary_print <- null_tar_summary_tbl %>%
  group_by(specimen, rname) %>%
  mutate(
    "Reads" = pNums(total.cnt, digits = 0),
    "Unchar." = pNums(unc.freq),
    "Complete" = pNums(com.freq),
    "Insertion" = pNums(in.freq),
    "Deletion" = pNums(del.freq)) %>%
  ungroup() %>%
  select(specimen, rname, Reads, Unchar., Complete, Insertion, Deletion) %>%
  gather(Type, val, -specimen, -rname) %>%
  spread(rname, val, fill = "-") %>%
  mutate(
    specimen = factor(
      specimen, 
      levels = unique(str_extract(sample_info$sampleName, "[\\w]+"))),
    Type = factor(
      Type , 
      levels = c("Reads", "Unchar.", "Complete", "Insertion", "Deletion"))) %>%
  arrange(specimen, Type) %>%
  rename(specimen = "Specimen") %>%
  mutate(Specimen = ifelse(
    duplicated(as.character(Specimen)), " ", as.character(Specimen)))

emphasize.strong.rows(which(!grepl(" ", null_tar_summary_print$Specimen)))
pander(
  null_tar_summary_print, caption = tbl_caps[5], 
  justify = "center", style = "simple", 
  col.names = pandoc.strong.return(names(null_tar_summary_print)))

```

```{r null_tar_graphic, fig.width=7.5, fig.height=2.5, fig.cap=fig_caps[7]}
null_tar_plot_data <- null_tar_summary_tbl %>%
  select(-total.cnt) %>%
  gather(key = type, value = freq, -specimen, -rname) %>%
  mutate(
    rname = factor(
      rname, levels = panel_targets$target[panel_targets$edit == "null"]),
    specimen = factor(
      specimen, levels = unique(str_extract(sample_info$sampleName, "[\\w]+"))),
    type = factor(
      type, levels = c("unc.freq", "com.freq", "in.freq", "del.freq"))) %>%
  arrange(desc(type))

levels(null_tar_plot_data$type) <- c("Unchar.", "Complete", "Insertion", "Deletion")
  
ggplot(null_tar_plot_data, aes(x = rname, y = freq, group = specimen)) + 
  geom_bar(stat = "identity", aes(fill = type)) + 
  facet_wrap(~ specimen, nrow = 1) +
  scale_fill_brewer(type = "qual", palette = "Paired") +
  labs(x = NULL, y = "Frequency", fill = NULL) +
  custom_theme +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r null_tar_del_profile, eval=FALSE, fig.width=7.5, fig.height=3.5, fig.cap=fig_caps[8]}
null_del_profile <- edit_grl$D[edit_grl$D$edit == "null"]
null_del_profile <- null_del_profile[
  rep(seq_along(null_del_profile), null_del_profile$count)]

plot_indel_cov(null_del_profile, norm_targets, resolution = 1L) + 
  scale_y_continuous(trans = "log10", labels = comma) + 
  scale_fill_brewer(type = "qual", palette = "Set1", direction = -1) + 
  custom_theme + 
  theme(legend.position = "None")
```

```{r null_tar_in_profile, eval=FALSE, fig.width=7.5, fig.height=3.5, fig.cap=fig_caps[9]}
null_in_profile <- edit_grl$I[edit_grl$I$edit == "null"]
start(null_in_profile) <- start(null_in_profile) - 
  as.integer(null_in_profile$in.len/2)
width(null_in_profile) <- null_in_profile$in.len
null_in_profile <- null_in_profile[
  rep(seq_along(null_in_profile), null_in_profile$count)]

plot_indel_cov(null_in_profile, norm_targets, resolution = 1L) + 
  scale_y_continuous(trans = "log10", labels = comma) + 
  scale_fill_brewer(type = "qual", palette = "Set1", direction = -1) + 
  custom_theme + 
  theme(legend.position = "None")
```

```{r discussion, eval=config$discussion$include, child=file.path(config$Install_Directory, config$discussion$path)}
```

```{r methods, eval=config$methods, child=file.path(code_dir, "methods.Rmd")}
```

```{r signature, eval=config$signature$include & config$format == "pdf", child=file.path(code_dir, "signature.Rmd")}
```
