# -*- mode: Snakemake -*-
# Sequence Alignment: BWA based

rule build_ref:
  input: ROOT_DIR + "/" + config["Panel_Path"]
  output: temp(ROOT_DIR + "/genomes/" + PANEL + ".tmp.fasta")
  params: 
    tool="tools/rscripts/gen_panal_refs.R",
    ref=config["RefGenome"]
  log:
    RUN_DIR + "/logs/panel.ref.log"
  shell:
    "Rscript {params.tool} {input} -o {output} -r {params.ref} > {log} 2>&1"

rule index_ref:
  input: ROOT_DIR + "/genomes/" + PANEL + ".tmp.fasta"
  output: 
    ref=ROOT_DIR + "/genomes/" + PANEL + ".fasta",
    aux1=ROOT_DIR + "/genomes/" + PANEL + ".amb",
    aux2=ROOT_DIR + "/genomes/" + PANEL + ".ann",
    aux3=ROOT_DIR + "/genomes/" + PANEL + ".bwt",
    aux4=ROOT_DIR + "/genomes/" + PANEL + ".pac",
    aux5=ROOT_DIR + "/genomes/" + PANEL + ".sa"
  params: ROOT_DIR + "/genomes/" + PANEL
  shell:
    """
    cp {input} {output.ref}
    bwa index -p {params} {output.ref}
    """

rule align_R1:
  input:
    seq=RUN_DIR + "/processData/{sample}.R1.fastq.gz",
    genome=ancient(ROOT_DIR + "/genomes/" + PANEL + ".fasta"),
    aux1=ancient(ROOT_DIR + "/genomes/" + PANEL + ".amb"),
    aux2=ancient(ROOT_DIR + "/genomes/" + PANEL + ".ann"),
    aux3=ancient(ROOT_DIR + "/genomes/" + PANEL + ".bwt"),
    aux4=ancient(ROOT_DIR + "/genomes/" + PANEL + ".pac"),
    aux5=ancient(ROOT_DIR + "/genomes/" + PANEL + ".sa")
  output:
    temp(RUN_DIR + "/processData/{sample}.R1.pre.unsorted.bam")
  params: 
    bwa=config["BWAparams"],
    index=ROOT_DIR + "/genomes/" + PANEL
  shell:
    "bwa mem {params.bwa} {params.index} {input.seq} | samtools view -b -o {output}"

rule sort_R1:
  input: RUN_DIR + "/processData/{sample}.R1.pre.unsorted.bam"
  output: temp(RUN_DIR + "/processData/{sample}.R1.pre.bam")
  shell: "samtools sort {input} -o {output}"

rule index_R1:
  input: RUN_DIR + "/processData/{sample}.R1.pre.bam"
  output: temp(RUN_DIR + "/processData/{sample}.R1.pre.bai")
  shell: "samtools index -b {input} {output}"

rule align_R2:
  input:
    seq=RUN_DIR + "/processData/{sample}.R2.fastq.gz",
    genome=ancient(ROOT_DIR + "/genomes/" + PANEL + ".fasta"),
    aux1=ancient(ROOT_DIR + "/genomes/" + PANEL + ".amb"),
    aux2=ancient(ROOT_DIR + "/genomes/" + PANEL + ".ann"),
    aux3=ancient(ROOT_DIR + "/genomes/" + PANEL + ".bwt"),
    aux4=ancient(ROOT_DIR + "/genomes/" + PANEL + ".pac"),
    aux5=ancient(ROOT_DIR + "/genomes/" + PANEL + ".sa")
  output:
    temp(RUN_DIR + "/processData/{sample}.R2.pre.unsorted.bam")
  params: 
    bwa=config["BWAparams"],
    index=ROOT_DIR + "/genomes/" + PANEL
  shell:
    "bwa mem {params.bwa} {params.index} {input.seq} | samtools view -b -o {output}"

rule sort_R2:
  input: RUN_DIR + "/processData/{sample}.R2.pre.unsorted.bam"
  output: temp(RUN_DIR + "/processData/{sample}.R2.pre.bam")
  shell: "samtools sort {input} -o {output}"

rule index_R2:
  input: RUN_DIR + "/processData/{sample}.R2.pre.bam"
  output: temp(RUN_DIR + "/processData/{sample}.R2.pre.bai")
  shell: "samtools index -b {input} {output}"

rule align_merged:
  input:
    seq=RUN_DIR + "/processData/{sample}.consol.fasta",
    genome=ancient(ROOT_DIR + "/genomes/" + PANEL + ".fasta"),
    aux1=ancient(ROOT_DIR + "/genomes/" + PANEL + ".amb"),
    aux2=ancient(ROOT_DIR + "/genomes/" + PANEL + ".ann"),
    aux3=ancient(ROOT_DIR + "/genomes/" + PANEL + ".bwt"),
    aux4=ancient(ROOT_DIR + "/genomes/" + PANEL + ".pac"),
    aux5=ancient(ROOT_DIR + "/genomes/" + PANEL + ".sa")
  output:
    temp(RUN_DIR + "/processData/{sample}.post.unsorted.bam")
  params: 
    bwa=config["BWAparams"],
    index=ROOT_DIR + "/genomes/" + PANEL
  shell:
    "bwa mem {params.bwa} {params.index} {input.seq} | samtools view -b -o {output}"
    
rule sort_bam:
  input: RUN_DIR + "/processData/{sample}.post.unsorted.bam"
  output: RUN_DIR + "/processData/{sample}.post.bam"
  shell: "samtools sort {input} -o {output}"

rule index_bam:
  input: RUN_DIR + "/processData/{sample}.post.bam"
  output: RUN_DIR + "/processData/{sample}.post.bai"
  shell: "samtools index -b {input} {output}"
